File: 1671-minimum-number-of-removals-to-make-mountain-array/1671-minimum-number-of-removals-to-make-mountain-array.py
class Solution:
    def minimumMountainRemovals(self, nums: List[int]) -> int:
        n = len(nums)
        lis = [1] * (n)
        lds = [1] * (n)
        maxBitonicLength = 1
        
        for idx in range (1, n):
            for prev in range (0, idx):
                if nums[prev] < nums[idx] and lis[idx] < lis[prev] + 1:
                    lis[idx] = lis[prev] + 1
        
        for idx in range (n - 2, -1, -1):
            for prev in range (n - 1, idx, -1):
                if nums[prev] < nums[idx] and lds[idx] < lds[prev] + 1:
                    lds[idx] = lds[prev] + 1
                    
        for idx in range (0, n):
            if lis[idx] > 1 and lds[idx] > 1:
                maxBitonicLength = max(maxBitonicLength, lis[idx] + lds[idx] - 1)
                
        return n - maxBitonicLength
                
                

File: 1671-minimum-number-of-removals-to-make-mountain-array/1671-minimum-number-of-removals-to-make-mountain-array.java
class Solution {
    public int minimumMountainRemovals(int[] nums) {
        int n = nums.length, maxBitonic = 0;
        int[] lis = new int[n];
        int[] lds = new int[n];
        for(int idx = 0; idx < n; ++idx) {
            lis[idx] = 1;
            lds[idx] = 1;
        }
        for(int idx = 1; idx < n; ++idx)
            for(int prev = 0; prev < idx; ++prev)
                if(nums[prev] < nums[idx] && lis[prev] + 1 > lis[idx])
                    lis[idx] = lis[prev] + 1;
        for(int idx = n - 2; idx >= 0; --idx)
            for(int prev = n - 1; prev > idx; --prev)
                if(nums[prev] < nums[idx] && lds[prev] + 1 > lds[idx])
                    lds[idx] = lds[prev] + 1;
        for(int idx = 0; idx < n; ++idx)
            if(lds[idx] > 1 && lis[idx] > 1)
                maxBitonic = Math.max(maxBitonic, lds[idx] + lis[idx] - 1);
        return (n - maxBitonic);
    }
}